font/src/ft/mod.rs

```diff
@@ -238,42 +256,25 @@ impl FreeTypeRasterizer {
         Ok(FullMetrics { size_metrics, cell_width: width })
     }
 
-    fn get_matching_face(
+    fn query_font(
         &mut self,
+        pattern: Pattern,
+        hash: PatternHash,
         desc: &FontDesc,
-        slant: Slant,
-        weight: Weight,
     ) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.set_weight(weight.into_fontconfig_type());
-        pattern.set_slant(slant.into_fontconfig_type());
-        pattern.add_pixelsize(self.pixel_size);
-
-        self.query_font(pattern, desc)
-    }
-
-    fn get_specific_face(&mut self, desc: &FontDesc, style: &str) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.add_style(style);
-        pattern.add_pixelsize(self.pixel_size);
-
-        self.query_font(pattern, desc)
-    }
-
-    fn query_font(&mut self, pattern: Pattern, desc: &FontDesc) -> Result<FontKey, Error> {
         let config = fc::Config::get_current();
-        let fonts = fc::font_sort(&config, &mut pattern.clone())
+        let patterns = fc::font_sort(&config, &mut pattern.clone())
             .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
 
-        let mut font_iter = fonts.into_iter();
+        let mut pattern_iter = patterns.into_iter();
 
-        let base_font = font_iter.next().ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
-        let base_font = pattern.render_prepare(config, base_font);
+        let base_pattern =
+            pattern_iter.next().ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        let base_pattern = pattern.render_prepare(config, base_pattern);
+        let base_hash = FontHash::new(hash, base_pattern.hash());
```

Comment:
Do we hash the unprepared and prepared pattern together here? That seems like a very strange choice. Is there any benefit to doing that here, or is it only relevant for fallback fonts?

Reply:
it's basically hashing of requested (size + family) and rendered thing, where size could be different to requested one.